source/rust_verify/src/names.rs

```diff
@@ -0,0 +1,338 @@
+    //   - rustc emits a library file on disk using its erase_ghost internal names
+    //       - in erase_ghost: impl%2
+    //   - we serialize and emit our VIR AST in CrateWithMetadata
+    //     - the serialized VIR AST contains impl_Vec7
+    //   - we also serialize the erase_ghost mapping in CrateWithMetadata
+    //     - ("impl", [2]) -> impl_Vec7
+    //     - see the export_impls function below and see import_export.rs
+    // - when a client application imports vstd:
+    //   - rustc reads in the erase_ghost-compiled vstd library from disk
+    //     - this file contains the *same* erase_ghost rustc internal names
+    //       created while compiling vstd
+    //       - in erase_ghost: impl%2
+    //     - we again represent these internal names with the same RustcLocalImplId
+    //       as we did while compiling vstd
+    //       - in erase_ghost: ("impl", [2])
+    //   - we read in the serialized mapping from CrateWithMetadata created when compiling vstd
+    //     - in erase_ghost: ("impl", [2]) -> impl_Vec7
+    //   - we use the RustcLocalImplId and deserialized map_to_stable_name to recover the same
+    //     Ident values that were used while compiling vstd
+    //     - impl_Vec7
+    //   - the key result is that these Ident values will match the !erase_ghost Ident values
+    //     that are present in the rest of the deserialized VIR AST for the vstd library
+    //     (whereas if we hadn't done the remapping, we would have seen inconsistent names,
+    //     impl%2 from the !erase_ghost rustc-emitted library, and impl%3 in our deserialized
+    //     VIR AST)
+    pub(crate) map_to_stable_name: HashMap<RustcLocalImplId, Ident>,
+}
+
+#[derive(Debug)]
+pub(crate) struct ImplNameState {
+    impl_table: HashMap<ImplFingerprint, Ident>,
+}
+
+fn def_path_to_rustc_id<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_path: DefPath,
+    num_segments: usize,
+) -> RustcLocalImplId {
+    let mut path = vec![tcx.crate_name(def_path.krate).to_string()];
+    let mut disambiguators: Vec<u32> = Vec::new();
+    let mut i: usize = 0;
+    for d in def_path.data.iter() {
+        use rustc_hir::definitions::DefPathData;
+        if i >= num_segments {
+            break;
+        }
+        match &d.data {
+            DefPathData::ValueNs(symbol) | DefPathData::TypeNs(symbol) => {
+                path.push(symbol.to_string());
+            }
+            DefPathData::Ctor => {
+                path.push(vir::def::RUST_DEF_CTOR.to_string());
+            }
+            DefPathData::Impl => {
+                // We don't have names for the impls at this point, so just use "impl"
+                path.push("impl".to_string());
+                disambiguators.push(d.disambiguator);
+            }
+            _ => {}
+        }
+        i += 1;
+    }
+    (path, disambiguators)
+}
+
+fn def_path_to_path<'tcx>(tcx: TyCtxt<'tcx>, def_path: DefPath) -> TypPath {
+    def_path_to_rustc_id(tcx, def_path, usize::MAX).0
```

Comment:
Can we add an assert here that the disambiguator list we're ignoring is empty?

Reply:
> Can we add an assert here that the disambiguator list we're ignoring is empty?

I just tried this:

```
fn def_path_to_path<'tcx>(tcx: TyCtxt<'tcx>, def_path: DefPath) -> TypPath {
    let (path, disambiguators) = def_path_to_rustc_id(tcx, def_path, usize::MAX);
    assert!(disambiguators.len() == 0);
    path
}
```

However, this assertion fails for nested impls (which, admittedly, we don't support yet, but we could in the future, and `names.rs` is intended to account for this):
```
struct S;
impl S {
    fn f() {
        struct Q;
        impl Q {
            fn g() {}
        }
    }
}
```

Reply:
Oh, I... had no idea nested impls were a thing.

Can you give a rough argument why the fingerprinting is sound even when disambiguators are thrown away?